modules/standard/IO.chpl

```diff
@@ -1623,6 +1624,70 @@ proc file.init(fp: c_FILE, hints=ioHintSet.empty) throws {
   initHelper(this, fp, hints);
 }
 
+private proc initHelper2(ref f: file, fd: c_int, hints = ioHintSet.empty,
+                         style:iostyleInternal = defaultIOStyleInternal()) throws {
+
+  var local_style = style;
+  f._home = here;
+  extern proc chpl_cnullfile():c_FILE;
+  var err = qio_file_init(f._file_internal, chpl_cnullfile(), fd, hints._internal, local_style, 0);
+
+  // On return, either f._file_internal.ref_cnt == 1, or f._file_internal is
+  // NULL.
+  // err should be nonzero in the latter case.
+  if err {
+    var path_cs:c_string;
+    var path_err = qio_file_path_for_fd(fd, path_cs);
+    var path = if path_err then "unknown"
+                           else createStringWithNewBuffer(path_cs,
+                                                          policy=decodePolicy.replace);
+    try ioerror(err, "in file.init", path);
+  }
+}
+
+@unstable "initializing a file with a style argument is unstable"
+proc file.init(fd: c_int, hints=ioHintSet.empty, style:iostyle) throws {
+  this.init();
+
+  initHelper2(this, fd, hints, style);
+}
+
+/*
+
+Create a Chapel file that works with a system file descriptor.  Note that once
+the file is open, you will need to use a :proc:`file.reader` or
+:proc:`file.writer` to create a channel to actually perform I/O operations
+
+.. note::
+
+  This is an alternative way to create a :record:`file`.  The main way to do so
+  is via the :proc:`open` function.
+
+The system file descriptor will be closed when the Chapel file is closed.
+
+.. note::
+
+  This function can be used to create Chapel files that refer to system file
+  descriptors that do not support the ``seek`` functionality. For example, file
+  descriptors that represent pipes or open socket connections have this
+  property. In that case, the resulting file value should only be used with one
+  :record:`fileReader` or :record:`fileWriter` at a time.
+  The I/O system will ignore the channel offsets when reading or writing
+  to files backed by non-seekable file descriptors.
+
+
+:arg fd: a system file descriptor.
+:arg hints: optional argument to specify any hints to the I/O system about
+            this file. See :record:`ioHintSet`.
+
+:throws SystemError: Thrown if the file descriptor could not be retrieved.
+*/
+proc file.init(fd: c_int, hints=ioHintSet.empty) throws {
```

Comment:
It seems like our file initializers (and generally most Chapeltastic interfaces) should take Chapel types rather than C types.  Can we change this to take an `int` argument and to `.safeCast()` it to `c_int` when it needs to pass it off to the thing (I'm guessing a C interface) that needs it to be a C integer?  Sorry if I missed this in previous design discussions—reviewing #20143, I'm not seeing the type of the argument discussed.

Reply:
It looks like you proposed it over on the issue about removing `fd_t`, but I missed including that in the updated proposal: https://github.com/chapel-lang/chapel/issues/20128

Reply:
But note that even with removing this case, we still have a file initializer that takes in a `c_FILE`.  Which is part of why there was a proposal for moving these functions into the OS module, but we ultimately decided against that

Reply:
These initializers are pretty low-level interfaces intended entirely for use with C interoperability, so I think there's justification for having an interface that takes a C type in this scenario.  And there doesn't seem to be a problem sending int literals to the c_int argument in the updates I've made

Reply:
I understand that there's a rationale for why the argument could justifiably be a `c_int`, but I don't understand why we would prefer it to be a `c_int` over an `int`.  For any other routine where we wrap a C routine, we typically want to take in Chapel types and convert them to the C types under the covers.  Why would we want to do differently for this initializer?

Reasons why we'd want to take an `int`:
* it's more Chapeltastic and more likely to be familiar to Chapel users
* it would support sending in `const` and `var` arguments that were `int`s rather than just `param`s
* there doesn't seem to be any significant downside